sympy/solvers/inequalities.py

```diff
@@ -982,3 +988,338 @@ def reduce_inequalities(inequalities, symbols=[]):
+            if atom.func == Le:
+                return self._assert_upper(sym, c)
+
+        raise ValueError(f"{atom} could not be asserted.")
+    def _assert_upper(self, xi, ci):
+        if ci >= self.upper[xi]:
+            return "SAT", self.assign
+        if ci < self.lower[xi]:
+            return "UNSAT", {xi >= self.lower[xi], xi <= ci}
+        self.upper[xi] = ci
+        if xi in self.nonslack and self.assign[xi] > ci:
+            self._update(xi, ci)
+
+        return "OK", None
+
+    def _assert_lower(self, xi, ci):
+        if ci <= self.lower[xi]:
+            return "SAT", self.assign
+        if ci > self.upper[xi]:
+            return "UNSAT", {xi <= self.upper[xi], xi >= ci}
+        self.lower[xi] = ci
+        if xi in self.nonslack and self.assign[xi] < ci:
+            self._update(xi, ci)
+
+        return "OK", None
+
+    def _update(self, xi, v):
+        i = self.col_index[xi]
+        for j, b in enumerate(self.slack):
+            aji = self.A[j, i]
+            self.assign[b] = self.assign[b] + aji*(v - self.assign[xi])
+        self.assign[xi] = v
+
+    def backtrack(self):
+        pass
+
+    def check(self):
+        def _debug_internal_state_printer1(iteration, A, bas, variables):
+            if not SYMPY_DEBUG:
+                return
+            import sys
+            from sympy.matrices.dense import Matrix
+            from sympy import pprint
+
+            bvar = [None]*len(bas)
+            for v, idx in bas.items():
+                bvar[idx] = v
+
+            r1 = Matrix([variables])
+            c1 = Matrix(bvar)
+            corner = Matrix([[iteration]])
+
+            tableau = Matrix([[corner, r1], [c1, A]])
+            pprint(tableau)
+            sys.stderr.write("\n")
+            sys.stderr.write(f"{self.assign}\n")
+            for v in self.upper:
+                sys.stderr.write(str(self.lower[v]) + " <= " + str(v) + " <= " + str(self.upper[v]) + "\n")
+
+        def _debug_internal_state_printer2(xi, xj):
+            if not SYMPY_DEBUG:
+                return
+            import sys
+            sys.stderr.write(f"\npivoting {xi} with {xj}\n\n")
+
+        from sympy.matrices.dense import Matrix
+        M = self.A.copy()
+        basic = {s: i for i, s in enumerate(self.slack)}  # contains the row index associated with each basic variable
+        nonbasic = set(self.nonslack)
+        iteration = 0
+        while True:
+            iteration += 1; _debug_internal_state_printer1(iteration, M, basic, self.all_var)
+
+            # nonbasic variables always must be within bounds
+            assert all(((self.assign[nb] >= self.lower[nb]) == True) and ((self.assign[nb] <= self.upper[nb]) == True) for nb in nonbasic)
+
+            # assignments for x must always satisfy Ax = 0
+            X = Matrix([self.assign[v] for v in self.col_index])
+            assert all(abs(val) < 10**(-10) for val in M*X) # TODO: Maybe return None if this cond can't be met?
```

Comment:
Maybe we can have a variable that's set when the tests are being run so that these assert statements are only run during the tests. Maybe pytest has a way to this.

Reply:
@oscarbenjamin any thoughts on how this could be done? Is there a variable that gets set when the tests are run?

Reply:
Why not have a parameter like `check=False` and then call the function with check=True during the test code? Then it is easier to enable checking in any situation rather than just the test suite.

The kind of thing that I would potentially want an assert like this for is to pick up on problems that might arise in cases that are *not* tested e.g. because user code does something unanticipated. If the asserts only run during testing then they can just be part of the test suite.